backface-visibility:hidden should create a stacking context and containing block.

mattwoodrow opened the pull request:
#### 2440e9efc74d6acbb62f9e024daa7f731e394759
<pre>
backface-visibility:hidden should create a stacking context and containing block.
<a href="https://bugs.webkit.org/show_bug.cgi?id=261532">https://bugs.webkit.org/show_bug.cgi?id=261532</a>
&lt;rdar://114732608&gt;

Reviewed by Simon Fraser.

As of <a href="https://github.com/w3c/csswg-drafts/pull/9348">https://github.com/w3c/csswg-drafts/pull/9348</a>, backface-visibility:hidden should create a containing block and stacking context, if the element participates in a 3D rendering context.

This adds a new preference for this change (currently disabled, until other engines are ready to ship the change), and makes the code changes.

It also removes old WPTs that relied on the previous behaviour, and replaces them with new ones testing the new spec.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002-expected.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003-expected.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003.tentative-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004-expected.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004.tentative-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005-expected.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005.tentative-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-animated-002-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-animated-002.html: Removed.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::updateFromStyle):
(WebCore::RenderBox::computeHasTransformRelatedProperty const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::clippingRootForPainting const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Canonical link: <a href="https://commits.webkit.org/267980@main">https://commits.webkit.org/267980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd3d3be41978f463a34f223ec76b3f8af9ddd7fd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18244 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18578 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20083 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17073 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18439 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18732 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18465 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18675 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/15885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20966 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15898 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/16638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23134 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/15807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16917 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/16809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21023 "Passed tests") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/17522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17371 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21601 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16470 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/16477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5294 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4345 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20833 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/22834 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17218 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5141 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -504,10 +504,27 @@ void RenderBox::updateFromStyle()
+    if (styleToUse.backfaceVisibility() != BackfaceVisibility::Hidden)
+        return false;
+
+    if (!element() || !element()->parentElement() || !element()->parentElement()->renderer())
```
Would be nice to avoid the repeated long chains of dereferencing.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2175,6 +2175,9 @@ RenderLayer* RenderLayer::clippingRootForPainting() const
         if (current->transform() || compositedWithOwnBackingStore(*current))
             return const_cast<RenderLayer*>(current);
 
+        if (renderer().settings().css3DTransformBackfaceVisibilityInteroperabilityEnabled() &&  current->participatesInPreserve3D() && current->renderer().style().backfaceVisibility() == BackfaceVisibility::Hidden)
```
Extra space after `&&`

smfr commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -889,6 +889,20 @@ CFNetworkNetworkLoaderEnabled:
     WebKit:
       default: false
 
+CSS3DTransformBackfaceVisibilityInteroperabilityEnabled:
```
Is there sufficient value in putting this behind a setting if it's immediately enabled?

mattwoodrow replied:
This was a copy-paste mistake :)

The intention is for it to be in 'preview' state, while the other engines also implement the change and we get a feel for any backward compatibility issues.

mattwoodrow committed:
backface-visibility:hidden should create a stacking context and containing block.
https://bugs.webkit.org/show_bug.cgi?id=261532
<rdar://114732608>

Reviewed by Simon Fraser.

As of https://github.com/w3c/csswg-drafts/pull/9348, backface-visibility:hidden should create a containing block and stacking context, if the element participates in a 3D rendering context.

This adds a new preference for this change (currently disabled, until other engines are ready to ship the change), and makes the code changes.

It also removes old WPTs that relied on the previous behaviour, and replaces them with new ones testing the new spec.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002-expected.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003-expected.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003.tentative-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-003.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004-expected.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004.tentative-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-004.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005-expected.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-002.tentative-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005.tentative-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-005.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-animated-002-expected.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/backface-visibility-hidden-animated-002.html: Removed.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::updateFromStyle):
(WebCore::RenderBox::computeHasTransformRelatedProperty const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::clippingRootForPainting const):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Canonical link: https://commits.webkit.org/267980@main
